rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,634 @@
+message AgeEncryptionKeyWithState {
+  // PublicKey is the Bech32 encoded age X25519 public key associated with the KeyState.
+  bytes public_key = 1;
+  // KeyState is the state PublicKey is currently in.
+  teleport.recordingencryption.v1.KeyState key_state = 2;
+}
+
+// GetRotationStateResponse contains the current KeyState of all WrappedKeys
+// in the EncryptedSessionRecordingConfig.
+message GetRotationStateResponse {
+  // Keys contains the list of currently active public keys along with their status.
+  repeated AgeEncryptionKeyKeyWithState keys = 1;
+}
+
+// CompleteRotationRequest
+message CompleteRotationRequest {}
+
+// CompleteRotationResponse
+message CompleteRotationResponse {}
+
+// UploadEncryptedRecordingRequest is an individual part of an encrypted session
+// recording .tar file.
+message UploadEncryptedRecordingRequest {
+  // SessionID the recording relates to.
+  string session_id = 1;
+  // PartIndex is the ordered index applied to the chunk.
+  int64 part_index = 2;
+  // Part is the encrypted part of session recording data being uploaded.
+  bytes part = 3;
+}
+
+// UploadEncryptedRecordingResponse
+message UploadEncryptedRecordingResponse {}
+```
+
+### Session Recording Modes
+
+There are four session recording modes that describe where the recordings are
+captured and how they're shipped to long term storage.
+
+- `proxy-sync`
+- `proxy`
+- `node-sync`
+- `node`
+  Where the recordings are collected is largely unimportant to the encryption
+  strategy, but whether or not they are handled async or sync has different
+  requirements.
+
+In sync modes the session recording data is written immediately to the auth
+service without intermediate disk writes. The auth service then handles
+batching and multipart upload of events to long term storage. In order to
+ensure resilience to partial uploads of parts, each batch will need to be
+encrypted individually on the auth service. This will likely mean wrapping the
+`gzipWriter` [used by slices](https://github.com/gravitational/teleport/blob/master/lib/events/stream.go#L669)
+with an `age` encryption writer.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to prevent all session recording
+data from being lost in the event of an agent crashing, each part will be
+encrypted individually much like the encrypted batches in `sync` modes. Because
+the data is already encrypted, exploding the final `.tar` file back into events
+to be uploaded to the auth service is not possible. Instead the auth service
+must accept a binary upload of the `.tar` file which it can then proxy to long
+term storage. This will be done using the `UploadEncryptedSessionRecording`
+streaming RPC. Each part will then be written to long term storage using ourt
+existing multipart upload functionality.
```

Comment:
Will the auth server check each part to confirm that when decrypted it contains valid events in the correct format?

Reply:
I haven't planned on doing this, but it could? I guess without this you could potentially upload arbitrary data provided it had a reasonable header 🤔